src/eddymotion/model/kernels.py

```diff
@@ -0,0 +1,174 @@
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+#
+import numpy as np
+from sklearn.gaussian_process.kernels import Hyperparameter, Kernel
+
+
+class SphericalCovarianceKernel(Kernel):
+    """
+    Custom kernel based on spherical covariance function.
+
+    Parameters
+    ----------
+    lambda_s : float, default=1.0
+        Scale parameter for the covariance function.
+    a : float, default=1.0
+        Distance parameter where the covariance function goes to zero.
+    sigma_sq : float, default=1.0
+        Noise variance term.
+    lambda_s_bounds : pair of floats >= 0 or "fixed", default=(1e-5, 1e4)
+        The lower and upper bound on lambda_s.
+    a_bounds : pair of floats >= 0 or "fixed", default=(1e-5, np.pi)
+        The lower and upper bound on a.
+    sigma_sq_bounds : pair of floats >= 0 or "fixed", default=(1e-5, 1e4).
+    """
+
+    def __init__(
+        self,
+        lambda_s=2.0,
+        a=0.1,
+        sigma_sq=1.0,
+        lambda_s_bounds=(1e-5, 1e4),
+        a_bounds=(1e-5, np.pi),
+        sigma_sq_bounds=(1e-5, 1e4),
+    ):
+        self.lambda_s = lambda_s
+        self.a = a
+        self.sigma_sq = sigma_sq
+        self.lambda_s_bounds = lambda_s_bounds
+        self.a_bounds = a_bounds
+        self.sigma_sq_bounds = sigma_sq_bounds
+
+    @property
+    def hyperparameter_lambda_s(self):
+        return Hyperparameter("lambda_s", "numeric", self.lambda_s_bounds)
+
+    @property
+    def hyperparameter_a(self):
+        return Hyperparameter("a", "numeric", self.a_bounds)
+
+    @property
+    def hyperparameter_sigma_sq(self):
+        return Hyperparameter("sigma_sq", "numeric", self.sigma_sq_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """
+        Compute the kernel matrix.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_samples)
+            Precomputed pairwise angles.
+        Y : array-like of shape (n_samples, n_samples), default=None
+            Second input for the kernel function.
+        eval_gradient : bool, default=False
+            Evaluate the gradient with respect to the log of
+            the kernel hyperparameter.
+
+        Returns
+        -------
+        K : array-like of shape (n_samples, n_samples)
+            Kernel matrix.
+        K_gradient : array-like of shape (n_samples, n_samples, n_dims), optional
+            The gradient of the kernel matrix with respect to the log of the
+            hyperparameters. Only returned when `eval_gradient` is True.
+        """
+        if theta_prime is not None:
+            theta = theta_prime
+        K = np.where(theta <= self.a, 1 - 3 * (theta / self.a) ** 2 + 2 * (theta / self.a) ** 3, 0)
+
+        K_gradient = None
+        if eval_gradient:
+            K_gradient = np.zeros((theta.shape[0], theta.shape[1], 3))
```

Comment:
This feels odd. X should be (n_samples, n_features), and for DWI this is (N, 3) with N gradient directions. I think the gradient differences indeed must be calculated inside the kernel.


```suggestion
            K_gradient = np.zeros((X.shape[0], X.shape[1], 3))
```